### ISSUE TITLE
Softmax applied twice ?

### ISSUE THREAD

**Issue by drauh** [icoxfog417/tying-wv-and-wc#1]
Hello,

I have come across your code looking for a way to tie weights in keras, and it is really useful thanks !

By looking at the AugmentedModel specification, especially augmented_loss function; it seems to me that you are applying softmax activation twice to your output, line 30 and line 33.

Malo

**Comment by icoxfog417** [comment]
Thank you for referring to my code!

It'll be fixed now. To calculate the loss, we need two softmax results.

* for **J_t**: Ordinally softmax is needed to compute the softmax_cross_entropy.
* for **J_aug**: The temperature based softmax to compute the augmented loss (KL distance)

![image](https://user-images.githubusercontent.com/544269/28866789-f602ce1a-77ae-11e7-8e69-44800d99e95a.png)

In the implementation, if we stack the *Activation Layer* then we can calculate **J_t** but can not **J_aug** (need inverse-calculation of softmax).  So I [remove the *Activation Layer*](https://github.com/icoxfog417/tying-wv-and-wc/blob/master/model/augmented_model.py#L61) and [compute softmax in the loss method](https://github.com/icoxfog417/tying-wv-and-wc/blob/master/model/augmented_model.py#L32).

**Comment by drauh** [comment]
Oh your right I missed that pop before the compilation !

I my implementation I still wanted to use the softmax activation layer so to get back the logits I applied a log (almost the inverse of the softmax). To my understanding this gives the same result because adding a constant doesn't change anything to the softmax.


        loss = K.categorical_crossentropy(y_pred, y_true)

        # y is (batch x seq x vocab)
        y_indexes = K.argmax(y_true, axis=2)  # turn one hot to index. (batch x seq)
        y_vectors = self.embedding(y_indexes)  # lookup the vector (batch x seq x vector_length)

        #v_length = self.setting.vector_length
        #y_vectors = K.reshape(y_vectors, (-1, v_length))
        #y_t = K.map_fn(lambda v: K.dot(self.embedding.embeddings, K.reshape(v, (-1, 1))), y_vectors)
        #y_t = K.squeeze(y_t, axis=2)  # unknown but necessary operation
        #y_t = K.reshape(y_t, (-1, self.sequence_size, self.vocab_size))

        # vector x embedding dot products (batch x seq x vocab)
        y_t = K.tf.tensordot(y_vectors, K.transpose(self.embedding.embeddings), 1)
        y_t.set_shape((None, self.sequence_size, self.vocab_size))  # explicitly set shape
        y_t = K.softmax(y_t / self.temperature)
        y_pred_t = K.softmax(K.log(y_pred) / self.temperature)
        aug_loss = kullback_leibler_divergence(y_t, y_pred_t)
        loss += (self.gamma * self.temperature) * aug_loss

Thanks again for sharing your work.

**Comment by drauh** [comment]
One another level, it is not clear for me in the paper how to handle the bias of the last layer. To calculate **J_t** they seemed to use them but not for calculating **J_aug**. This seems to be a little painful to implement. 

Maybe by using use_bias=False in the projection layer, then a regularization layer as in [there](https://github.com/fchollet/keras/issues/5563) to calculate **J_aug** and then applying a 'bias' layer (which is not in keras but implementing it is easy you just have to remove the kernel things from the dense layer).

If your are interested in this, maybe I can try to implement that for you.

**Comment by icoxfog417** [comment]
thank you for your pointing out. Indeed the bias is not applied to the augmented loss.

**for J_t & ouput**

![image](https://user-images.githubusercontent.com/544269/28963838-8ccdb5c0-7945-11e7-9967-1ca505823901.png)

**for J_aug**

![image](https://user-images.githubusercontent.com/544269/28963851-96535bae-7945-11e7-9efe-1eb151e21de4.png)

In my implementation, the bias is not considered (and as you say, it will be a little hard to do it in Keras). 
So I'm very glad if you share your challenge!

(Sometimes I feel I want to touch TensorFlow directly...)
